test: Cypress | Flaky fixes + Replacing static waits

Aishwarya-U-R opened the pull request:
## Description
- This PR includes the flaky fixes of the below specs:
           - jsEditor.NavigateToNewJSEditor() using TS helpers (Addressing flakyness in specs /BugTests/JS_Bug29131_spec.ts, BugTests/invalidLintError_Spec.ts, ClientSide/BugTests/JS_Bug28764_Spec.ts, /ClientSide/BugTests/Bug29566_Spec.ts, ServerSide/JsFunctionExecution/JSFunctionExecution_spec.ts)
           - /Widgets/ListV2/Listv2_onItemClick_spec.js (toast validation updated to ts helper)
           - /Apps/MongoDBShoppingCart_spec.ts (Flaky fix 2nd case)
           - /Regression/Apps/PromisesApp_spec.js (1st test)
           - /ClientSide/OtherUIFeatures/ApplicationURL_spec.js (whole spec for static wait removal + flaky fixes)
           - /ClientSide/Templates/Fork_Template_spec.js (complete spec fixes, cypress exception fix)
           - /Sanity/Datasources/Arango_Basic_Spec.ts (3rd test)
           - /ServerSide/OnLoadTests/JSOnLoad2_Spec.ts (8th case flaky fix)
           - /ClientSide/MobileResponsiveTests/AutoFillWidgets_Reflow_spec.ts (2nd testcase, flow update)
           - /ClientSide/OtherUIFeatures/Resize_spec.js (import update)
           - /ClientSide/Templates/Fork_Template_spec.js (1st & 3rd for flaky fix, removed 2nd - redundant check)
           - /Git/GitSync/RepoLimitExceededErrorModal_spec.js flaky fix for EE repo failure
           - /ServerSide/ApiTests/API_Bugs_Spec.js (3rd case, removed waits, moved to TED Api)
           - /Apps/EchoApiCMS_spec.js (removed redundant action)
           - /AppNavigation/Sidebar_spec.ts, /AppNavigation/TopInline_spec.ts, /AppNavigation/TopStacked_spec.ts - added refresh calls for CI flakyness
- homePage.RenameWorkspace() & homePage.NavigateToHome() & homePage.Signout() - removed static wait, added dynamic check
- homePage.CreateNewApplication() - removed static wait
- homePage.CreateAppInWorkspace() - removed commented code
- homePage.CreateNewWorkspace() - logic improved
- agHelper.AssertURL() - added timeout, removed static wait, calling assertHelper.AssertDocumentReady()
- Few of agHelper methods timeout & error mesg text update
- cy.Signup() - updated validations
- agHelper.CheckForErrorToast() to agHelper.FailIfErrorToast() - method name
- /ClientSide/BugTests/DSDiscardBugs_spec.ts - removed unnecessary method call

#### Type of change
- Script fix (non-breaking change which fixes an issue)

## Testing
#### How Has This Been Tested?
- [X] Cypress CI runs

## Checklist:
#### QA activity:
- [X] Added `Test Plan Approved` label after all changes were reviewed

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Refactor**
  - Enhanced timeout handling across various helper methods for improved stability.
  - Streamlined interactions with UI elements using updated helper methods.
  - Adjusted test suites to utilize configuration values for consistency.

- **Tests**
  - Temporarily disabled specific test suites to address current issues.
  - Updated test assertions and interaction patterns for clarity and reliability.
  - Added new test cases and assertions to cover additional functionalities.

- **Chores**
  - Updated workflow variables for more efficient CI processes.
  - Curated the list of limited tests to focus on critical test scenarios.

- **Documentation**
  - Clarified test descriptions and intentions for better understanding.

- **Bug Fixes**
  - Fixed UI element interactions to prevent test flakiness and improve user experience.
  - Implemented page refresh before application imports to ensure clean state for tests.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7420655729

Aishwarya-U-R commented:
/ci-test-limit runId=7420655729

Aishwarya-U-R commented:
/ci-test-limit runId=7421557462

Aishwarya-U-R commented:
/ci-test-limit runId=7421557462

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7444442250

Aishwarya-U-R commented:
/ci-test-limit runId=7444442250

Aishwarya-U-R commented:
/ci-test-limit runId=7444442250

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7444442250

Aishwarya-U-R commented:
/ci-test-limit runId=7444968656

Aishwarya-U-R commented:
/ci-test-limit runId=7444968656

Aishwarya-U-R commented:
/ci-test-limit runId=7444968656

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7444968656

Aishwarya-U-R commented:
/ci-test-limit runId=7444968656

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit runId=7445994970

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7457219692

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit runId=7458073410

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7463985142

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit runId=7465639913

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit runId=7465813092

Aishwarya-U-R commented:
/ci-test-limit

Aishwarya-U-R commented:
/ci-test-limit

mohanarpit commented on app/client/cypress/e2e/Regression/ClientSide/OtherUIFeatures/ApplicationURL_spec.js:
```diff
@@ -14,21 +16,19 @@ describe("Slug URLs", () => {
+      cy.visit(`/applications/${applicationId}/pages/${pageId}/edit`, {
+        timeout: Cypress.config().pageLoadTimeout,
+      }).then(() => {
+        cy.wait(5000);
```
@Aishwarya-U-R Any reason we are waiting for 5 secs here? Can this be avoided?

Aishwarya-U-R replied:
Yes, it was 10 secs before. Without wait, the cy.visit of the new url is not fully loaded & test fails. Tried adding other dynamic assert, could not get the case to pass. Hence left it, so could ask help from others to take a look too.

mohanarpit commented on app/client/cypress/e2e/Regression/ClientSide/OtherUIFeatures/ApplicationURL_spec.js:
```diff
@@ -135,9 +132,8 @@ describe("Slug URLs", () => {
+      homePage.Signout(true);
       agHelper.VisitNAssert(url + "?embed=true&a=b", "signUpLogin");
-      agHelper.Sleep(2000);
       // cy.location().should((loc) => {
```
Please remove the commented code here as well.

Aishwarya-U-R replied:
Thanks for pointing out. Done.

mohanarpit commented on app/client/cypress/e2e/Regression/ClientSide/Templates/Fork_Template_spec.js:
```diff
@@ -7,49 +7,45 @@ describe(
       _.templates.SwitchToTemplatesTab();
 
       _.agHelper.GetNClick(templateLocators.templateCard);
+      _.agHelper.CheckForErrorToast("INTERNAL_SERVER_ERROR");
```
Sorry, I don't understand this validation. Why are we asserting that a toast message saying 'internal server error' should show up? Am I reading this wrong?

Aishwarya-U-R replied:
Here, we are verifying that if INTERNAL_SERVER_ERROR is thrown by Product during CI run & error is thrown in tc if it does:
 ```
public CheckForErrorToast(error: string) {
    cy.get("body").then(($ele) => {
      if ($ele.find(this.locator._toastMsg).length) {
        if ($ele.find(this.locator._specificToast(error)).length) {
          throw new Error("Error Toast from Application:" + error);
        }
      }
    });
  }
```

mohanarpit replied:
Sorry, I'm slightly more confused now. As per the function `CheckForErrorToast`, if Cypress finds the toast message which matches the error string, it'll error out. This means that the assertion will fail in the test case. Am I reading this code correctly?

Aishwarya-U-R replied:
Yes, assertion will fail the test case. Thus bringing out the Internal Server Error thrown by Product. The reason for having this check here is to catch such toast thrown by Product randomly.

mohanarpit replied:
But the product shouldn't throw "Internal server error". That's always a bug! Why are we encoding a bug in the test case? This test should fail in the scenario that the product throws an internal server error.

Aishwarya-U-R replied:
The test will fail if its caught during CI run. **Thats the whole point of having this check here.** As we have seen product silently Throwing Internal server error toasts randomly in CI runs.

mohanarpit replied:
Damn. I just got it. Can we please rename the function to `failIfErrorToast` ? The function name makes it sound like we are asserting the existence of the toast message. 

Sorry for the confusion here.

Aishwarya-U-R replied:
Right, done.

mohanarpit commented on app/client/cypress/e2e/Regression/ServerSide/OnLoadTests/JSOnLoad2_Spec.ts:
```diff
@@ -193,9 +193,9 @@ describe(
         myFun2: async () => {	},
         myFun3: async () => {	},
         myFun4: async () => {	},
-        myFun5: async () => {	},
```
Can you help me understand this change? Why are we replacing `myFun5` with `myFun8`?

Aishwarya-U-R replied:
Unable to get the screenshot of the run now, but the failure in this spec was something like JSObj.func5 already exists. Though, I could not replicate the failure in local runs. Hence updated the function name.

mohanarpit commented on app/client/cypress/support/Pages/AggregateHelper.ts:
```diff
@@ -485,7 +485,7 @@ export class AggregateHelper {
   public WaitUntilEleDisappear(selector: string) {
     cy.waitUntil(() => this.GetElement(selector, "not.exist"), {
-      errorMsg: "Element did not disappear even after 10 seconds",
+      errorMsg: "Element did not disappear even after 60 seconds",
```
Is this error message correct? The interval set in this function seems to be 1 sec.

mohanarpit commented on app/client/cypress/support/Pages/AggregateHelper.ts:
```diff
@@ -485,7 +485,7 @@ export class AggregateHelper {
   public WaitUntilEleDisappear(selector: string) {
     cy.waitUntil(() => this.GetElement(selector, "not.exist"), {
-      errorMsg: "Element did not disappear even after 10 seconds",
+      errorMsg: "Element did not disappear even after 60 seconds",
```
Possible to have this error message print: `'Element did not disappear after ' + Cypress.config().pageLoadTimeout + seconds` ? This way if the value of the config changes, the error message will be accurate. Else, we will see Cypress logs print 60 secs but the actual config value may be lesser.

Aishwarya-U-R replied:
Got it, done.

mohanarpit commented on app/client/cypress/support/Pages/AggregateHelper.ts:
```diff
@@ -913,14 +912,14 @@ export class AggregateHelper {
     let shouldFocus = true;
     let parseSpecialCharSeq = false;
-    let delay = 5;
+    let delay = 10;
```
I'm assuming this is the delay between typing 2 chars. If yes, any reason we are increasing the delay here?

Aishwarya-U-R replied:
10 is default delay for typing text by [Cypress](https://docs.cypress.io/api/commands/type#:~:text=considered%20animating.-,delay,-10). There seems to be some chars missing when typing, when using with delay : 5 leading to flakiness when trying to rename workspaces. Hence given it default value now.

mohanarpit commented on app/client/cypress/support/Pages/AggregateHelper.ts:
```diff
@@ -1606,8 +1605,11 @@ export class AggregateHelper {
+      "include",
+      url,
+    );
+    this.Sleep(); //new url to open
```
Any reason we left this static wait here? Can the function return after `cy.url` function call since we are already asserting there?

Aishwarya-U-R replied:
I think further steps were failing if this will be removed. I will check this by removing in next PR.

Aishwarya-U-R commented:
/ok-to-test

Aishwarya-U-R commented:
/ok-to-test tags="@tag.All"

sarojsarab commented on app/client/cypress/support/Pages/AggregateHelper.ts:
```diff
@@ -1606,8 +1614,11 @@ export class AggregateHelper {
+      "include",
+      url,
+    );
+    this.Sleep(); //new url to open
```
Can we wait until the document is ready instead of static wait and try once?

Aishwarya-U-R replied:
Done

sarojsarab commented on app/client/cypress/e2e/Regression/ClientSide/OtherUIFeatures/ApplicationURL_spec.js:
```diff
@@ -14,21 +16,19 @@ describe("Slug URLs", () => {
+      cy.visit(`/applications/${applicationId}/pages/${pageId}/edit`, {
+        timeout: Cypress.config().pageLoadTimeout,
+      }).then(() => {
+        cy.wait(5000);
```
Can we try to wait until document load here and check once?

Aishwarya-U-R replied:
Done

Aishwarya-U-R commented:
@sarojsarab - These are failing because this [run is limited run](https://github.com/appsmithorg/appsmith/pull/30067#issuecomment-1886908588). Addressed your comments. Can you please review now?
